tests/testsuite/artifact_dep.rs

```diff
@@ -0,0 +1,2049 @@
+                name = "foo"
+                version = "0.0.0"
+                authors = []
+                resolver = "2"
+                
+                [build-dependencies]
+                bar = { path = "bar/", artifact = ["bin", "staticlib", "cdylib"] }
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .file("build.rs", r#"
+            fn main() {
+                let baz: std::path::PathBuf = std::env::var("CARGO_BIN_FILE_BAR_baz").expect("CARGO_BIN_FILE_BAR_baz").into();
+                println!("{}", baz.display());
+                assert!(&baz.is_file()); 
+                
+                let lib: std::path::PathBuf = std::env::var("CARGO_STATICLIB_FILE_BAR_bar").expect("CARGO_STATICLIB_FILE_BAR_bar").into();
+                println!("{}", lib.display());
+                assert!(&lib.is_file()); 
+                
+                let lib: std::path::PathBuf = std::env::var("CARGO_CDYLIB_FILE_BAR_bar").expect("CARGO_CDYLIB_FILE_BAR_bar").into();
+                println!("{}", lib.display());
+                assert!(&lib.is_file()); 
+                
+                let dir: std::path::PathBuf = std::env::var("CARGO_BIN_DIR_BAR").expect("CARGO_BIN_DIR_BAR").into();
+                println!("{}", dir.display());
+                assert!(dir.is_dir());
+                
+                let bar: std::path::PathBuf = std::env::var("CARGO_BIN_FILE_BAR").expect("CARGO_BIN_FILE_BAR").into();
+                println!("{}", bar.display());
+                assert!(&bar.is_file()); 
+                
+                let bar2: std::path::PathBuf = std::env::var("CARGO_BIN_FILE_BAR_bar").expect("CARGO_BIN_FILE_BAR_bar").into();
+                println!("{}", bar2.display());
+                assert_eq!(bar, bar2);
+            }
+        "#)
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [package]
+                name = "bar"
+                version = "0.5.0"
+                authors = []
+                
+                [lib]
+                crate-type = ["staticlib", "cdylib"]
+            "#,
+        )
+        // compilation target is native for build scripts unless overridden
+        .file("bar/src/bin/bar.rs", &format!(r#"fn main() {{ assert_eq!(std::env::var("TARGET").unwrap(), "{}"); }}"#, cross_compile::native()))
+        .file("bar/src/bin/baz.rs", "fn main() {}")
+        .file("bar/src/lib.rs", "")
+        .build();
+    p.cargo("build -Z unstable-options -Z bindeps")
+        .masquerade_as_nightly_cargo()
+        .with_stderr_contains("[COMPILING] foo [..]")
+        .with_stderr_contains("[COMPILING] bar v0.5.0 ([CWD]/bar)")
+        .with_stderr_contains("[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]")
+        .run();
+
+    let build_script_output = build_script_output_string(&p, "foo");
+    let msg = "we need the binary directory for this artifact along with all binary paths";
+    #[cfg(any(not(windows), target_env = "gnu"))]
```

Comment:
These "gnu" cfg expressions are a bit hard for me to parse.  Can it maybe be simplified to be:

```rust
if cfg!(target_env = "msvc") {
     // ...msvc check goes here
} else {
    // ...not msvc goes here
}
```

I think something like that can make it clearer that the msvc target is the odd one out.

Reply:
That's so much better, and I think it should work. CI will be the judge of f84309f79be6d089faa4437f06aaeae22cf4fcdd , and I am optimistic.